### PR TITLE
[lldb][AIX] Added XCOFF Object File Header for AIX

### PR DESCRIPTION
**Pull request by DhruvSrivastavaX**
This PR is in reference to porting LLDB on AIX.

Link to discussions on llvm discourse and github:
1. https://discourse.llvm.org/t/port-lldb-to-ibm-aix/80640
2. #101657
The complete changes for porting are present in this draft PR:
#102601

Added XCOFF Object File Header for AIX.

Added base functionality for XCOFF support. Will enhance the files in incremental PRs

Details about XCOFF file format on AIX: [XCOFF](https://www.ibm.com/docs/en/aix/7.3?topic=formats-xcoff-object-file-format)

Review Request: @DavidSpickett

### CODE REVIEW

**Comment by DhruvSrivastavaX** [comment]
Hi Community Members, 

Apologies for being away for a while as I was occupied with some other activities. 
Here's some update at what all we are working on based on the past discussions:

1. Removed the CMAKE flag __AIX__ and replaced it with the LLVM environment's _AIX flag [**Done**]

2. We are working on merging some of the common files (like HostInfoAIX.cpp etc) derived from Linux while avoiding any major impact.
3. Working on creating a draft PR as suggested by @DavidSpickett , to highlight all the #if (_AIX) kind of code in the draft PR, to see the overall impact. 
4. Implementing additional functionalities which are currently missing on AIX.
5. Rearranging some of the code in the draft PR based on the previous discussions. 

Going forward @Lakshmi-Surekha will also be assisting me in some of the upstreaming activities. 

Thanks & Regards

**Comment by DhruvSrivastavaX** [comment]
> A couple of quick notes:
> 
> * are you sure that ObjectFileELF is the best thing to start with here? Based on the name, I would expect that ObjectFilePECOFF would be a better baseline. Or maybe the object file format is different enough from all of the existing ones that this kind of comparison is not useful?
> * without also seeing the cpp file, it hard to judge how similar it is to the other plugin
> * instead of redeclaring all the header types, would it be possible to use some structures from llvm? (I know that ObjectFileELF does not do that, but that's because it is very old. ObjectFilePECOFF -- a much newer class -- does just that)
> * a lot of the object file code should be testable on its own. If you include the necessary plugin glue in this PR, then you should be able to write tests similar to those in `lldb/test/Shell/ObjectFile`. For example, if you could start with a test similar to `test/Shell/ObjectFile/ELF/basic-info.yaml` and then include enough of code in the PR to make that pass.

Hi @labath , 

1. Actually there might be slight similarities with ObjectFilePECOFF, but the overall format structure is very different to be compared to any one of the existing formats, so I have just taken the ObjectFileELF.h as a base to follow the general class/file arrangement. 
2. As per our last discussion, I have planned to drop each of the major files separately, so I can push the .cpp file as another commit, and both of these can be used for comparison. 
3. From a surface level I can see that ObjectFilePECOFF has also created its own new structures based on its file format, essentially we have done something similar for XCOFF. Although, let me check the existing llvm structures for XCOFF thoroughly and get back to you on this. 
4. So, as per your suggestion, what path flow should we take? 
Based on previous discussions, as of now I am planning a commit for .h file, another commit for .cpp file and a final commit for the test cases. Is that okay?

**Comment by labath** [comment]
> > A couple of quick notes:
> > 
> > * are you sure that ObjectFileELF is the best thing to start with here? Based on the name, I would expect that ObjectFilePECOFF would be a better baseline. Or maybe the object file format is different enough from all of the existing ones that this kind of comparison is not useful?
> > * without also seeing the cpp file, it hard to judge how similar it is to the other plugin
> > * instead of redeclaring all the header types, would it be possible to use some structures from llvm? (I know that ObjectFileELF does not do that, but that's because it is very old. ObjectFilePECOFF -- a much newer class -- does just that)
> > * a lot of the object file code should be testable on its own. If you include the necessary plugin glue in this PR, then you should be able to write tests similar to those in `lldb/test/Shell/ObjectFile`. For example, if you could start with a test similar to `test/Shell/ObjectFile/ELF/basic-info.yaml` and then include enough of code in the PR to make that pass.
> 
> Hi @labath ,
> 
>     1. Actually there might be slight similarities with ObjectFilePECOFF, but the overall format structure is very different to be compared to any one of the existing formats, so I have just taken the ObjectFileELF.h only as a base to follow the general class/file arrangement, is all.
> 
>     2. As per our last discussion, I have planned to drop each of the major files separately, so I can push the .cpp file as another PR, and both of these can be used for comparison.

That's okay, though I fear you're taking this a bit too literally. I definitely wanted to split that PR up, but this is a bit too far. It would be better to have the header and the matching cpp file in the same PR, since you usually need to look at them in tandem.

That said, the main reason we (or I, at least) wanted to see this kind of a PR is to determine whether it makes sense to merge this with some of the existing code. For ObjectFileELF, I'm pretty confident the answer to that is going to be "no" (I'm slightly less sure about ObjectFilePECOFF, but I think I'd be willing to take your word for it -- I think a good litmus test would be whether the PECOFF and XCOFF parsing code in llvm shares anything). Given that, the comparison to a different object file plugin is not really interesting to me, as I believe that means this part of code should be treated as any new code that's being added to llvm. In particular, that it should be subject to the [incremental development](https://llvm.org/docs/DeveloperPolicy.html#incremental-development) policy.

This is where I was coming from when I wrote the previous message (I did not include the context, as I was in a bit of a 
hurry -- sorry). Incremental development means developing (and reviewing, and submitting) the change in small pieces that can be easily reviewed and come with their own tests. What exactly is a small piece is not easy to determine, but it's usually less than "an entire file". For an object file plugin, a relatively reasonable sequence would be:
- parse the file headers (enough so you can detect the file format and answer questions like what is its architecture)
- parse sections
- parse symbols
- other stuff (?)

I know this is going to be more work, and it's going to be a bit awkward to re-engineer a finished file into an incremental set of changes, but what you have is essentially a long-term development branch -- exactly the thing which the policy discourages. I'm not asking this because I want to be difficult, but large changes **really** are hard to review.

**Comment by DhruvSrivastavaX** [comment]
Okay sure. 
Thats not a problem Pavel, better to have it in smaller understandable pieces, so that it can be organised and reviewed properly. 

There are a couple of incoming code with such huge changes,  it will be good to adhere to this systematic approach. 

So, I will go ahead and modify this PR to drop a smaller piece of ObjectFileXCOFF.h and ObjectFileXCOFF.cpp instead, 
which can be reviewed easily (some base skeleton and declaration of parent virtual functions etc). 
Once that gets pulled in, I will build my next few PRs on that in order to complete these new additions. 
Will also try to add relevant test cases. 

Thanks!

**Comment by labath** [comment]
Thank you for your understanding.

Before trying to split up the other large files (a fairly large undertaking), I'd still recommend doing the initial thing of dropping the file in a PR (which should be relatively fast) and getting some feedback. ObjectFile plugins are one of the (few) things which I know that can be reasonably split and tested in multiple PRs. Unfortunately, most of the other plugins are much trickier, so we may need to get more creative.

**Comment by DhruvSrivastavaX** [comment]
Hi @labath , 
I have dropped some base structure and yaml test cases for the ObjectFileXCOFF support. 
Please provide your comments.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  PluginManager::RegisterPlugin(GetPluginNameStatic(),
+                                GetPluginDescriptionStatic(), CreateInstance,
+                                CreateMemoryInstance, GetModuleSpecifications);
+}
+
+void ObjectFileXCOFF::Terminate() {
+  PluginManager::UnregisterPlugin(CreateInstance);
+}
+
+bool UGLY_FLAG_FOR_AIX __attribute__((weak)) = false;
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:68]: We're going to have to find a different way to achieve whatever this is trying to achieve. Let's remove that, as it should not be necessary for this patch.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ObjectFileXCOFF.h"
+
+#include <algorithm>
+#include <cassert>
+#include <string.h>
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:14]:
```suggestion
#include <cstring>
```

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+#include <unordered_map>
+
+#include "lldb/Core/Module.h"
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:17]: Are all of these needed for the thing that the file is doing? (Please don't reduce them too aggressively -- that leads to fragile headers -- but I don't think you're using e.g. LZMA or PointerUnion at the moment)

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+ByteOrder ObjectFileXCOFF::GetByteOrder() const { return eByteOrderBig; }
+
+bool ObjectFileXCOFF::IsExecutable() const { return true; }
+
+uint32_t ObjectFileXCOFF::GetAddressByteSize() const { return 8; }
+
+lldb::SymbolType
+ObjectFileXCOFF::MapSymbolType(llvm::object::SymbolRef::Type sym_type) {
+  return lldb::eSymbolTypeInvalid;
+}
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:164]: looks like this isn't used (yet).

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  MapFileDataWritable(const lldb_private::FileSpec &file, uint64_t Size,
+                      uint64_t Offset);
+
+  uint32_t ParseDependentModules();
+
+private:
+  std::unique_ptr<llvm::object::XCOFFObjectFile> m_binary;
+  lldb_private::Address m_entry_point_address;
+  std::optional<lldb_private::FileSpecList> m_deps_filespec;
+  std::map<std::string, std::vector<std::string>> m_deps_base_members;
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:113]: unused, please remove for now.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+UUID ObjectFileXCOFF::GetUUID() { return UUID(); }
+
+uint32_t ObjectFileXCOFF::ParseDependentModules() {
+  ModuleSP module_sp(GetModule());
+  if (!module_sp)
+    return 0;
+
+  std::lock_guard<std::recursive_mutex> guard(module_sp->GetMutex());
+  if (m_deps_filespec)
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:188]: I'm confused. Who fills in this variable?

**Review by labath — COMMENTED** [review]
This is exactly what I had in mind -- thank you.

The patch looks mostly good, apart from the inline comments. My main question is about the "dependant modules" parsing code. AFAICT, its not actually functional (and not tested). Is yaml2objs xcoff backend sufficiently developed to create a test for the dependant module functionality? Given that we rely on llvm for parsing that, we don't have to test it extensively, but it would be nice to have at least one test. We could either do that in this patch, or rip out the dependant module functionality, and add it back in another patch -- up to you.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp (hunk 1 above)
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp]: Yes, sure. It is actually used in other files as a flag. But we can check on it later.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp (hunk 3 above)
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:17]: Filtered some of the headers. Thanks!

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp (hunk 6 above)
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp]: Filtered out for now.

**Comment by DhruvSrivastavaX** [comment]
> This is exactly what I had in mind -- thank you.
> 
> The patch looks mostly good, apart from the inline comments. My main question is about the "dependant modules" parsing code. AFAICT, its not actually functional (and not tested). Is yaml2objs xcoff backend sufficiently developed to create a test for the dependant module functionality? Given that we rely on llvm for parsing that, we don't have to test it extensively, but it would be nice to have at least one test. We could either do that in this patch, or rip out the dependant module functionality, and add it back in another patch -- up to you.

Great. Thanks for your inputs too. 
Yes, I think we can add dependant modules related changes later. I have removed them for now.
About the yaml xcoff testing, 
For a basic test case as this one, it works great! I also tried using obj2yaml and on a compiled executable for a basic .c test case, and then reversing it with yaml2obj and that works fine too. But for the dependant module functionality, I might need to check some more, we can check on that in the later PRs.

On another note, I have a doubt or maybe I am missing something: I am able to run lldb-test successfully on my aix system setup for this new basic xcoff file (where I have the entire aix changes along with a reduced xcoff same as in this PR). 
But lldb-test does not recognize the object file generated by yaml2obj on a linux ppc64 system (where I have only the merged changes and the xcoff changes only present in this PR). Is it not supposed to work that way or am I missing something?
Any suggestions about what can I try would be great!

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp (hunk 1 above)
Comment (labath) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp]: I see its used in the dwarf parser. We can deal with that later. My suggestion would be to look at how the llvm dwarf parser (llvm-dwarfdump) does this.

**Comment by labath** [comment]
> But lldb-test does not recognize the object file generated by yaml2obj on a linux ppc64 system (where I have only the merged changes and the xcoff changes only present in this PR). Is it not supposed to work that way or am I missing something?

It should work. It's likely that you missed some critical (small) piece of functionality that causes lldb to think the input is not valid. It should be fairly easy to step through the relevant code in `lldb-test` and see where it bails out (then to implement whatever is necessary to make it not do that). We don't run lldb tests on ppc64 on a regular basis, so it's also possible you're missing hitting some bug specific to that host, but I don't think that's the (only) problem here, as the test doesn't pass for me on x86 either.

**Comment by DhruvSrivastavaX** [comment]
Ok right. 
Yes, and it is working fine on AIX.
Let me get a little more familiar with lldb-test's working and I will try to figure out what's wrong.
Thanks!

**Comment by DhruvSrivastavaX** [comment]
Ok right. 
Yes, and it is working fine on AIX.
Let me get a little more familiar with lldb-test's working and I will try to figure out what's wrong.
Thanks!

**Comment by DhruvSrivastavaX** [comment]
Hi @labath, 
So the error was due to the endianness difference between Linux and AIX. 
The previous failure with the new test case has gone now.
I have dropped a fix so that AIX binary is recognised on little endian platforms as well, but
if there is any better way to add this check, please do let me know. 
Thanks!

**Comment by labath** [comment]
> Hi @labath, So the error was due to the endianness difference between Linux and AIX. The previous failure with the new test case has gone now. I have dropped a fix so that AIX binary is recognised on little endian platforms as well, but if there is any better way to add this check, please do let me know. Thanks!

Are XCOFF files always big endian? If so, then all you need is to add  a `DataExtractor::SetByteOrder()`. If not (and you want to support both endiannesses), then it would be better to add both cases to `XCOFFHeaderSizeFromMagic`.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+bool ObjectFileXCOFF::MagicBytesMatch(DataBufferSP &data_sp,
+                                      lldb::addr_t data_offset,
+                                      lldb::addr_t data_length) {
+  lldb_private::DataExtractor data;
+  data.SetData(data_sp, data_offset, data_length);
+  lldb::offset_t offset = 0;
+  uint16_t magic = data.GetU16(&offset);
+  if (magic == 0xF701)
+    magic = 0x01F7; /* Since AIX is big endian, and the host checking platform
+                       might be little endian. */
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:137]: I.e., something like this

```suggestion
  DataExtractor data(data_sp, data_offset, data_length);
  data.SetByteOrder(eByteOrderBig);
  lldb::offset_t offset = 0;
  uint16_t magic = data.GetU16(&offset);
```

Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp]: Great. Done, Thanks!

**Comment by DhruvSrivastavaX** [comment]
> Are XCOFF files always big endian? If so, then all you need is to add a `DataExtractor::SetByteOrder()`. If not (and you want to support both endiannesses), then it would be better to add both cases to `XCOFFHeaderSizeFromMagic`.

Yes, AIX XCOFF does not support little endian, so I went ahead with the SetByteOrder Fix.

**Review by labath — APPROVED** [review]
Looks good. Thanks. Do you need someone to push the merge button?

**Comment by DhruvSrivastavaX** [comment]
> Looks good. Thanks. Do you need someone to push the merge button?

Ok, great. Yes, please merge it, I don't have the permission to do that yet.

**Pull request merged**